ignite/distributed/utils.py

```diff
@@ -0,0 +1,220 @@
+            slurm_port = slurm_port[-4:]
+            os.environ["MASTER_PORT"] = str(int(slurm_port) + 15000)
+            # master address is the first hostname of nodes list
+            hostnames = subprocess.check_output(["scontrol", "show", "hostnames", os.environ["SLURM_JOB_NODELIST"]])
+            os.environ["MASTER_ADDR"] = hostnames.split()[0].decode("utf-8")
+        else:
+            os.environ["RANK"] = os.environ.get("RANK", "0")
+            os.environ["LOCAL_RANK"] = os.environ.get("LOCAL_RANK", "0")
+            os.environ["WORLD_SIZE"] = os.environ.get("WORLD_SIZE", "1")
+            os.environ["MASTER_ADDR"] = os.environ.get("MASTER_ADDR", "127.0.0.1")
+            os.environ["MASTER_PORT"] = os.environ.get("MASTER_PORT", "15000")
+        self._master_port = int(os.environ["MASTER_PORT"])
+        # for debug
+        self._master_addr = os.environ["MASTER_ADDR"]
+        self._local_rank = int(os.environ["LOCAL_RANK"])
+
+    def get_local_rank(self) -> int:
+        assert dist.is_available() and dist.is_initialized()
+        return self._local_rank
+
+    @staticmethod
+    def get_rank() -> int:
+        assert dist.is_available() and dist.is_initialized()
+        return dist.get_rank()
+
+    @staticmethod
+    def get_world_size() -> int:
+        assert dist.is_available() and dist.is_initialized()
+        return dist.get_world_size()
+
+    @staticmethod
+    def is_distributed() -> bool:
+        return dist.is_available() and dist.is_initialized() and dist.get_world_size() > 0
+
+    @staticmethod
+    def name() -> str:
+        return "dist"
+
+
+# default is sequential model
+_model = SequentialModel()
+# default backend is None
+_backend = None
+# default device is None
+_device = None
+# the model, device and backend could change at initialization
+
+
+def is_dist_distributed() -> bool:
+    return _model.is_distributed() and _model.name() != "xla"
+
+
+def is_tpu_distributed() -> bool:
+    return _model.is_distributed() and _model.name() == "xla"
+
+
+def get_world_size() -> int:
+    return _model.get_world_size()
+
+
+def get_rank() -> int:
+    return _model.get_rank()
+
+
+def get_local_rank() -> int:
+    return _model.get_local_rank()
+
+
+def get_device():
+    return _device
+
+
+def get_backend():
+    return _backend
+
+
+def get_hostname() -> str:
+    return socket.gethostname()
+
+
+def _find_best_dist_backend():
+    assert dist.is_available()
+    # select best backend nccl first, gloo second
+    if dist.is_nccl_available():
+        return "nccl"
+    elif dist.is_gloo_available():
+        return "gloo"
+    else:
+        raise Exception("No backend found (missing gloo and nccl")
+
+
+def _find_best_dist_device():
+    # select best device cuda is possible otherwise cpu
+    if torch.cuda.is_available() and torch.cuda.device_count() > 0:
+        return "cuda"
+    else:
+        return "cpu"
+
+
+def initialize(backend=None, device=None):
```

Comment:
Do we really need to specify `device` ?
Seems like we have 
```
nccl <-> torch.cuda.current_device()
gloo <-> cpu
xla <-> xm.xla_device()
```

Reply:
I agree. I did it because I tested gloo + cuda.